cli/azd/design-notes/environments.md

```diff
@@ -0,0 +1,271 @@
+
+```bicep
+@description('The Azure location to use for all resources.')
+@allowed([
+    'westus2'
+    'centralus'
+])
+@metadata({
+    azd: {
+        type: 'location'
+    }
+})
+param location string
+```
+
+To start, we support the following types:
+
+`location`: The value should be the short name of an Azure location. All locations are supported by default, but the exiting `@allowed` may restrict this set.
+`azd-deployment-context`: The value should be an `azd` deployment context object. This is a JSON object that contains information about the project which may be useful for values when deploying. It will be described in further detail later.
+
+### Deployment Outputs
+
+When a deployment completes, the outputs from the deployment are stored in the `config.json` file under the "outputs" key (a sibling of "properties"). As we did on inputs, we re-use the schema of an outputs property of an ARM Deployment to save ourselves the trouble of inventing a new format.  When supporting Pulumi and Terraform we'll have to build an adapter but the format should be flexible enough to support this.
+
+### Environment Files
+
+An explicit design goal is that all environment state is stored in single JSON file, which should provide flexibility in developing some longer term story around collaboration.
+
+While no longer technically required, we continue produce a `.env` and `.parameters.json` in the `.azure\<environment-name>` folder use by external tooling. These are generated artifacts that are updated any time the authoritative values are updated in `config.json`.
+
+The generated `.env` file no longer contains deployment parameters.
+
+Now it is just the union of outputs from the deployment and any keys managed manually via `azd env set`. Since this file is regenerated each time a deployment happens, if a user removes an output from their `.bicep` file, the corresponding output will be removed from the `.env` file. 
+
+### ARM Deployments
+
+When using Bicep or ARM, deployments happen by creating an ARM Deployment object, which has a name that makes up part of it's identity. In addition, when preforming a subscription level deployment (the only form `azd` currently supports) you need to provide a location for where the deployment metadata is stored. When we support targeting resource groups, instead of a location we'll need a resource group name.
+
+For now, we'll introduce a `target` property in the environment:
+
+```json
+{
+  "id": "9fdfa455-2131-4faf-b2e3-8380aba97c96",
+  "name": "default",
+  "infra": {
+    "main": { 
+      "target": {
+        "type": "arm",
```

Comment:
Instead of "discriminating union" with "type" property, consider having separate objects for ARM, Terraform, Pulumi etc